File: JavaScript/Basics/Variables.js
/**
 * In JavaScript there is three different keywords you can use to define variables.
 * These keywords are "let", "var" and "const". I personally use "const" for most.
 */

/** 
 * The first keyword we will look at is "const". This keyword is used to declare a
 * constant block-scoped variable. If you use "const" you cannot change the value
 * of it through reassignment.
 * 
 * Documentation - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const
 * 
 * In the example below we will create a new constant called "hello".
 */

const hello = "Hello World!";
console.log(hello); // This will log - hello

hello = "Hello Exxon!"; // This will give us an error - "TypeError: Assignment to a constant variable."

/**
 * The second keyword we will look at is "let". This keyword is used to declare block-scoped
 * variables which can be reassigned later on.
 * 
 * Documentation - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let
 * 
 * In the example below we will create a new variable called "myName". We can then reassign 
 * the value of this variable later on.
 */

let myName = "Exxon";
console.log(myName); // This will log - Exxon

myName = "Tom";
console.log(myName); // This will log - Tom

/**
 * The final keyword we can look at is "var". This keyword is used to declare globally-scopped
 * variables.
 * 
 * Documentation - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var
 * 
 * In the example below we will create a new variable called "myAge". We can also reassign the
 * value of this variable later on, just like we did when using "let".
 */

var myAge = 19;
console.log(myAge); // This will log - 19

myAge = 18;
console.log(myAge); // This will log - 18

/**
 * The difference between "let" and "var".
 * 
 * The main difference is that "let" is a block-scoped variable and "var" is a globally-scoped"
 * variable. This means that "let" variables can only be used in the scope they are defined in.
 * "var" variables can be used globally, meaning they are not restricted to the scope they are
 * defined in. (I'm not the best at describing this)
 * 
 * In the example below we will define 4 variables. We will define "foo" and "bar" in the
 * global-scope and we will define "Foo" and "Bar" inside a block.
 */

var foo = "foo";
let bar = "bar";
console.log(foo, bar); // This will log - foo bar

{
    var Foo = "Foo";
    let Bar = "Bar";
    console.log(Foo, Bar); // This will log - Foo Bar
}

console.log(Foo); // This will log - Foo
console.log(Bar); // This will give us an error - "ReferenceError: Bar is not defined"

File: Python/Basics/Variables.py
""" 
When creating new variables in Python you do not need to use a command/keyword.
A variable is created the moment you first assign a value to it. You can
also change the type and the value of a variable after it has been set.

When creating a new variable there is certain names that you cannot have. You
cannot have a space or (-) in the variable name. The name also cannot start with
a number, it must start with a letter or an underscore (_). Variable names are
case-sensitive meaning that you can have a variable called "name" and another called
"Name".

Documentation - https://www.w3schools.com/python/python_variables.asp

In the example below we are going to create two variables, "myName" and "myAge".
We will then re-assign the value of these variables.
"""

myName = "Exxon"
myAge = 18

print(myName) # This will log - Exxon
print(myAge) # This will log - 18

myName = "Tom"
myAge = "19" 

print(myName) # This will log - Tom
print(myAge) # This will log - 19

# Notice how you do not get any errors by reassigning the value or the type of the variable.


"""
We can also asign values to multiple variables in one line. This can be done by listing the
variable names seperated by a comma (,) and then an equals sign and the values seperated by
a comma (,).

Documentation - https://www.w3schools.com/python/python_variables_multiple.asp

In the example below we will create three new variables, "hello", "foo" and "bar".
"""

hello, foo, bar = "Hello World!", "foo", "bar"

print(hello) # This will log - Hello World!
print(foo) # This will log - foo
print(bar) # This will log - bar


"""
We can also assin the same value to multiple variables in one line. This is done by listing 
each variable name seperated by " = " followed by an equals sign and then the value of the 
variables.

In the example below we will create two new variables, "exxon" and "Exxon".
"""

exxon = Exxon = "My name is Exxon!"

print(exxon) # This will log - My name is Exxon!
print(Exxon) # This will log - My name is Exxon!


"""
The final thing we will look at is unpacking a collection. If we have a collection of values in
a list we can extract the values into variables. This is called "unpacking".

In the example below we will have an array of fruits that we will unpack to three variables, "x",
"y", and "z".
"""

fruits = ["apple", "banana", "cherry"]
x, y, z = fruits

print(x) # This will log - apple
print(y) # This will log - banana
print(z) # This will log - cheery